pkg/client/unversioned/conditions.go

```diff
@@ -17,11 +17,69 @@ limitations under the License.
 package unversioned
 
 import (
+	"time"
+
 	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/errors"
 	"k8s.io/kubernetes/pkg/apis/extensions"
 	"k8s.io/kubernetes/pkg/util/wait"
 )
 
+// DefaultRetry is the recommended retry for a conflict where multiple clients
+// are making changes to the same resource.
+var DefaultRetry = wait.Backoff{
+	Steps:    5,
+	Duration: 10 * time.Millisecond,
+	Factor:   1.0,
+	Jitter:   0.1,
+}
+
+// DefaultBackoff is the recommended backoff for a conflict where a client
+// may be attempting to make an unrelated modification to a resource under
+// active management by one or more controllers.
+var DefaultBackoff = wait.Backoff{
+	Steps:    4,
+	Duration: 10 * time.Millisecond,
+	Factor:   5.0,
+	Jitter:   0.1,
+}
+
+// RetryConflict executes the provided function repeatedly, retrying if the server returns a conflicting
+// write. Callers should preserve previous executions if they wish to retry changes. It performs an
+// exponential backoff.
+//
+//     var pod *api.Pod
+//     err := RetryOnConflict(DefaultBackoff, func() (err error) {
+//       pod, err = c.Pods("mynamespace").UpdateStatus(podStatus)
+//       return
+//     })
+//     if err != nil {
+//       // may be conflict if max retries were hit
+//       return err
+//     }
+//     ...
+//
+// TODO: Make Backoff an interface?
+func RetryOnConflict(backoff wait.Backoff, fn func() error) error {
```

Comment:
The cockroachdb guys had a good set of utilities for this, I need to review
what they did and crib more closely.  Also, David noticed and I agreed that
the back off/poll callback function is more useful than retry on conflict.
I need to see whether I can find a happier refactoring.  Hold slightly.

On Oct 28, 2015, at 7:18 PM, Daniel Smith <notifications@github.com> wrote:

In pkg/client/unversioned/conditions.go
<https://github.com/kubernetes/kubernetes/pull/16067#discussion_r43331645>:

> +// write. Callers should preserve previous executions if they wish to retry changes. It performs an
> +// exponential backoff.
> +//
> +//     var pod *api.Pod
> +//     err := RetryOnConflict(DefaultBackoff, func() (err error) {
> +//       pod, err = c.Pods("mynamespace").UpdateStatus(podStatus)
> +//       return
> +//     })
> +//     if err != nil {
> +//       // may be conflict if max retries were hit
> +//       return err
> +//     }
> +//     ...
> +//
> +// TODO: Make Backoff an interface?
> +func RetryOnConflict(backoff wait.Backoff, fn func() error) error {

Minor, but I would consider making Retry a method of Backoff, so that users
write client.DefaultBackoff.RetryOnConflict(fn) instead of
client.RetryOnConflict(client.DefaultBackoff,
fn).

Hm, I guess that doesn't work when backoff is from another package.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/16067/files#r43331645>.